NEPT-1999: Schedule publishing of new version of node

enriquelacoma opened the pull request:
## NEPT-1999

### Description

Users need to be able to schedule the publishing of a new version of an already published node.
### Change log

- Added: add patches to the deploy 
    .- patches/scheduler_workbench-allowed_status.patch
    .- https://www.drupal.org/files/issues/scheduler_workbench-revision_publish-2048999-45.patch

Commit:
NEPT-1999: Apply patch to schedule revisions to publish and unpublish

Commit:
NEPT-1999: Add description to steps methods

Commit:
NEPT-1999: Update behat test tags

Commit:
NEPT-1999: Allow publication of revisions with selected status

Commit:
NEPT-1999: Include the patch on the platform patches

Commit:
NEPT-1999: Update behat test

imanoleguskiza committed:
Merge branch 'release-2.5' into nept-1999-release-2.5

imanoleguskiza commented on resources/multisite_drupal_standard.make:
```diff
@@ -656,6 +656,13 @@ projects[scheduler][version] = 1.5
+; https://webgate.ec.europa.eu/CITnet/jira/browse/NEPT-1999
+; https://www.drupal.org/project/scheduler_workbench/issues/2048999
+projects[scheduler_workbench][patch][] = https://www.drupal.org/files/issues/scheduler_workbench-revision_publish-2048999-45.patch
+; Allow publication of revisions with selected status. This patch requires the from issue https://www.drupal.org/project/scheduler_workbench/issues/2048999
```
There is a type here, it should say "This patch requires the patch from issue "

imanoleguskiza commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,147 @@
++    '#title' => t('Allowed Workbench Moderation status for scheduling publication'),
++    '#options' => $options,
++    '#default_value' => variable_get('scheduler_workbench_allowed_status', array()),
++    '#description' => t('Choose status to allow revisions for scheduling publication. When Scheduler Workbench tries to publish a revision i will check if the current state is one of the selected from this list.'),
```
There is a type here, it should say "When Scheduler Workbench tries to publish a revision **it** will check"

imanoleguskiza commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,147 @@
++    'description' => 'Select the states that can be scheduled for publication',
++    'page callback' => 'drupal_get_form',
++    'page arguments' => array('scheduler_workbench_allowed_status_form'),
++    'access arguments' => array('administer scheduler workbench'),
```
The permission name here differs from the one defined below: here its "administer scheduler workbench" but then we defined it as "scheduler workbench"

imanoleguskiza commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,147 @@
++/**
++ * Sets the allowed status that can be scheduled for publication.
++ */
++function _scheduler_workbench_status() {
```
The functione name here is a bit weird, it doesn't tell anything about what it does. I think a better name would be _scheduler_workbench_define_allowed_status

imanoleguskiza commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,147 @@
++  $status = array();
++  foreach ($transitions as $key => $value) {
++    if ($value->to_name != 'published') {
++      unset($transitions[$key]);
```
This interaction seems unnecessary, can't we simply add $value->from_name to the $status array if the condition is met?
`if ($value->to_name == 'published') {
  $status[] = $value->from_name;
}
`

imanoleguskiza commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,147 @@
+  */
+ function _scheduler_workbench_scheduler_pre_publish(&$node) {
++  // Get status allowed to be published.
++  $status = variable_get('scheduler_workbench_allowed_status', array());
```
The variable name is misleading here, I think $allowed_status is much better ( same for other appearances of the same variable)

imanoleguskiza commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,147 @@
++  // Get status allowed to be published.
++  $status = variable_get('scheduler_workbench_allowed_status', array());
++  foreach ($status as $key => $value) {
++    if ($value === 0) {
```
In which situation will value equal 0?

imanoleguskiza commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,147 @@
++    watchdog("scheduler_workbench", t("Node %node can't be published, check status allowed to be published."), $options, WATCHDOG_INFO);
++    return;
++  }
+   if ($state = variable_get('scheduler_publish_moderation_state_' . $node->type, FALSE)) {
```
The code added above should probable be inside this condition, that way we don't do unnecessary checks if the content type doesn't have scheduler enabled.

imanoleguskiza commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,147 @@
++    }
++  }
++  // Check if the revision current status allows it to be published.
++  if (!in_array($node->workbench_moderation['my_revision']->state, $status)) {
```
The check here is fine but I think we shouldn't allow the users to define a scheduling date if the revision state is not allowed to begin with.
I would do this on scheduler_workbench_form_alter, just before calling scheduler_workbench_add_revision_publish_on_field and I would add a message to inform the user that publishing the revision in that state is not allowed.

enriquelacoma replied:
This allows, for example, to schedule the publication of a draft and publish it only if it is in Validation, that way you can schedule a publication and publish it only if it is in a correct status.

If a revision is not in a valid status, i will not be published and register a watchdog message with the error.

I think that we could keep the current funtionality and, as you suggest, add a message informing the user that it will no be published with the current status.

Commit:
NEPT-1999: Fix typos, permission name in access check, improve functions
and variable names and improve conditions

Commit:
NEPT-1999: Add message when scheduling the publication if the revision
status does not alllowd it to be published

Commit:
Merge branch 'nept-1999-release-2.5' of https://github.com/ec-europa/platform-dev into nept-1999-release-2.5

Commit:
NEPT-1990: Fix condition to shows message

Commit:
NEPT-1999: Show message on validation

imanoleguskiza commented on tests/src/Context/SchedulerContext.php:
```diff
@@ -39,4 +58,33 @@ public function iChangeUnpublishingDateOfTypeNodeWithTitleToDate($type, $title,
+   *
+   * @Then I fill in :arg1 with current day
+   */
+  public function iFillInWithCurrentDay($arg1) {
```
This two steps seem to be unused so I guess we should delete them.

enriquelacoma replied:
We use them now in the scenario to check the publication of revision

imanoleguskiza commented on tests/features/scheduler.feature:
```diff
@@ -10,28 +10,32 @@ Feature: Scheduler features
   Scenario: User can schedule a date to publish a content
     When I go to "node/add/page"
```
The tests are fine but we are missing the tests where we publish a new revision of an existing node. We also need a tests that shows how a node with an incorrect state is not published when the date comes.

enriquelacoma replied:
Added two new scenarios

imanoleguskiza commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,173 @@
++        '%status' => $node->workbench_moderation['my_revision']->state,
++      );
++      watchdog("scheduler_workbench", t("Node %node can't be published, check status allowed to be published."), $options, WATCHDOG_INFO);
++      return;
```
In my tests the watchdog was triggered but the scheduling proceeded as usual so I think the check needs to be triggered somewhere else.

Commit:
NEPT-1999: Check status before node publishing

gillesdeudon committed:
Merge branch 'release-2.5' into nept-1999-release-2.5

Commit:
NEPT-1999: Add background configuration

Commit:
NEPT-1999: Add revision status to watchdog message and change the cron
step to avoid a problem with the value of
scheduler_workbench_allowed_status after updating allowed status

Commit:
NEPT-1999: Add scenarios to test the case for using a not allowed status
and the publication of a revision

enriquelacoma committed:
Merge branch 'release-2.5' into nept-1999-release-2.5

Commit:
NEPT-1999: Use current revision statusNEPT-1999: Use current revision
status

enriquelacoma committed:
NEPT-1999: Add comment to the local patch included in the make file

gillesdeudon commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,206 @@
++/**
++ * Implements hook_update_N().
++ */
++function scheduler_workbench_update_7001() {
```
We should not have this new hook_update in 2.5. I suggest to remove it from the "local" patch.

enriquelacoma replied:
Function removed

gillesdeudon commented on resources/multisite_drupal_standard.make:
```diff
@@ -653,6 +653,15 @@ projects[scheduler][version] = 1.5
+; Allow publication of revisions with selected status. This patch requires the patch from issue https://www.drupal.org/project/scheduler_workbench/issues/2048999
+; https://webgate.ec.europa.eu/CITnet/jira/browse/NEPT-1999
+; This patch requieres the patch https://www.drupal.org/files/issues/scheduler_workbench-revision_publish-2048999-45.patch from
+; this issue https://www.drupal.org/project/scheduler_workbench/issues/2048999.
```
We need a reference to the D.O issue from where this "local" patch id derived like that we can ensure the traceability of the patch and ease our job during future module upgrades.

enriquelacoma replied:
Comment updated

gillesdeudon reviewed (changes requested):
@enriquelacoma could you tackle the comments I made?

enriquelacoma committed:
NEPT-1999: Add reference to the drupal issue we reference

enriquelacoma committed:
NEPT-1999: Add current status to message and remove hokk_update_N

gillesdeudon commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,197 @@
++ * Implements hook_menu().
++ */
++function scheduler_workbench_menu() {
++  $items['admin/config/content/scheduler_workbench'] = array(
```
IMO, this menu time should be either a child of "Scheduler" item (admin/config/scheduler) or of "Workbench moderation" item (admin/config/workbench/moderation) in order to be consistent with the other contrib modules that enrich Scheduler or Workbench moderation like "Workbench moderation" does with the "Workbench" module. 

I am more in favour of "Scheduler" item with in the `scheduler_workbench_allowed_status_form()`, a mention of the Workbench moderation transitions and the admin form managing them.

enriquelacoma replied:
Included as a scheduler tab

gillesdeudon commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,197 @@
+-
++  if (!empty($node->revision_publish_on) || !empty($node->publish_on) ) {
++    // Check if the current status will be published.
++    $allowed_status = variable_get('scheduler_workbench_allowed_status', array());
```
IMO, if `scheduler_workbench_allowed_status`is not set, we should come back to the transition settings and only give the list of states that can be lead to "published"; instead of an empty array.

enriquelacoma replied:
Created wrapper _scheduler_workbench_allowed_status()

gillesdeudon commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,197 @@
+ function _scheduler_workbench_scheduler_pre_publish(&$node) {
+   if ($state = variable_get('scheduler_publish_moderation_state_' . $node->type, FALSE)) {
++    // Get status allowed to be published.
++    $allowed_status = variable_get('scheduler_workbench_allowed_status', array());
```
Same comment as I did for `scheduler_workbench_node_validate()`.

enriquelacoma replied:
Created wrapper _scheduler_workbench_allowed_status()

gillesdeudon commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,197 @@
+   workbench_moderation_moderate($current, $state);
+   $create_publishing_revision = variable_get('scheduler_publish_revision_' . $current->type, 0) == 1;
++  $allowed_status = variable_get('scheduler_workbench_allowed_status', array());
++  if (!in_array($current->workbench_moderation['my_revision']->state, $allowed_status)) {
```
Same remark as I made for `scheduler_workbench_node_validate()`. Maybe encapsulating this variable_get call in its own function could be interesting.

enriquelacoma replied:
Created wrapper _scheduler_workbench_allowed_status()

gillesdeudon commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,197 @@
++ * @see system_settings_form()
++ */
++function scheduler_workbench_allowed_status_form($form, &$form_state) {
++  $options = workbench_moderation_state_labels();
```
We should tie the options on the Workbench moderation transitions possibilities like done in `_scheduler_workbench_define_allowed_status()`. 
That means this function should move to the module file. I think also it could be integrated in the module API and become public.

enriquelacoma replied:
Created wrapper _scheduler_workbench_allowed_status()

gillesdeudon commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,197 @@
++    '#title' => t('Allowed Workbench Moderation status for scheduling publication'),
++    '#options' => $options,
++    '#default_value' => variable_get('scheduler_workbench_allowed_status', array()),
++    '#description' => t('Choose status to allow revisions for scheduling publication. When Scheduler Workbench tries to publish a revision it will check if the current state is one of the selected from this list.'),
```
In link with my comment above, we could add in the description a text referring to the "Transition" interface of Workbench moderation to add more Moderate state.

enriquelacoma replied:
Updated description to include default values

gillesdeudon commented on resources/patches/scheduler_workbench-allowed_status.patch:
```diff
@@ -0,0 +1,197 @@
+     if ($permitted_states = workbench_moderation_states_next($current_state, $draft_owner, $node)) {
+       $allowed = isset($permitted_states[$published_state]);
++    } // Get status allowed to be published.
++    $allowed_status = variable_get('scheduler_workbench_allowed_status', array());
```
Same remark as I made for scheduler_workbench_node_validate(). Maybe encapsulating this variable_get call in its own function could be interesting.

enriquelacoma replied:
Created wrapper _scheduler_workbench_allowed_status()

gillesdeudon reviewed (changes requested):
I leave some comments to address before being able to merge.

enriquelacoma committed:
NEPT-1999: Create wrapper function for scheduler_workbench_allowed_status default
value and include configuration page as a scheduler tab

enriquelacoma committed:
NEPt-1999: Update scheduler_workbench patch so the publis on and
unpublish on fields, have the same order on edition and on creation

ventzie555 reviewed (dismissed):
I tested it and everything appears to work fine. Code also look good to me.

Fefaine committed:
Merge branch 'release-2.5' into nept-1999-release-2.5

enriquelacoma committed:
NEPT-1999: On install, accept the scheduling of states that can be published and update tests

Commit:
NEPT-1999: Update readme

Commit:
NEPT-1999: Add default permissions for contributor

Commit:
NEPT-1999: Add behat test for contributor role, configure contributor
default permissions, show publication date for new nodes on moderate tab

Commit:
NEPT-1999: Show publication date on the moderate tab and behat test

Commit:
NEPT-1999: Remove unneded permiisions

Commit:
NEPT-1999: Add permission to unschedule for contributor, remove status
check and update behat test

AlbaParrilla committed:
Merge branch 'release-2.5' into nept-1999-release-2.5

aritomelo committed:
NEPT-1999: Setting user timezone as default timezone.

gillesdeudon committed:
Merge branch 'release-2.5' into nept-1999-release-2.5

aritomelo committed:
NEPT-1999: Fix typo on patch URL.

Commit:
Merge branch 'release-2.5' into nept-1999-release-2.5

enriquelacoma committed:
Merge branch 'release-2.5' into nept-1999-release-2.5

Commit:
Merge branch 'nept-1999-release-2.5' of https://github.com/ec-europa/platform-dev into nept-1999-release-2.5

Commit:
NEPT-1999: Fix conding standars

gillesdeudon committed:
Merge branch 'release-2.5' into nept-1999-release-2.5

gillesdeudon committed:
Merge branch 'release-2.5' into nept-1999-release-2.5

ku-enza committed:
Merge branch 'release-2.5' into nept-1999-release-2.5

Commit:
NEPT-1999: Published revisions are not scheduled

enriquelacoma committed:
Merge branch 'release-2.5' into nept-1999-release-2.5

gillesdeudon commented:
I approve in order to unblock the PR as it is in a new code review process

enriquelacoma committed:
Merge branch 'release-2.5' into nept-1999-release-2.5

enriquelacoma committed:
Merge branch 'release-2.5' into nept-1999-release-2.5

AlbaParrilla committed:
NEPT-1999: Modify message.

AlbaParrilla committed:
NEPT-1999: Fix message in behat test.

AlbaParrilla committed:
NEPT-1999: Fix message in behat test.

AlbaParrilla reviewed (approved):
I just changed a message a little bit, the rest is working properly. 
Ok for me

Fefaine reviewed (approved):
OK

Fefaine commented:
🎆
